#UC vs dNME or dMML
plot_dt=rbind(
  data.table(value=GR_merge$dNME[GR_merge$UC>=0.5],statistic="dNME"),
  data.table(value=GR_merge$dMML[GR_merge$UC>=0.5],statistic="dMML"))

ggplot(plot_dt,aes(x=value,fill=statistic))+geom_density(alpha=0.5)
plot_dt=rbind(
  data.table(UC=GR_merge$UC[GR_merge$dNME>=0.5],statistic="high dNME"),
  data.table(UC=GR_merge$UC[GR_merge$dMML>=0.5],statistic="high dMML"))

ggplot(plot_dt,aes(x=value,fill=statistic))+geom_density(alpha=0.5)

#Figure S1
digits_round=2
#Fig 1A:dNME vs dMML
GR_merge_dt=data.table(dMML=GR_merge$dMML,dNME=GR_merge$dNME,dNME_pval=GR_merge$dNME_pval,dMML_pval=GR_merge$dMML_pval)
GR_merge_dt=GR_merge_dt[GR_merge_dt$dNME_pval<=pval_cutoff|GR_merge_dt$dMML_pval<=pval_cutoff]
GR_merge_dt_agg=GR_merge_dt[, list(dNME=round(median(dNME),digits=digits_round),
                                   Bottom25=round(quantile(dNME,probs=0.25),digits=digits_round),
                                   top25=round(quantile(dNME,probs=0.75),digits=digits_round)), 
                            by = list(dMML = round(dMML,digits=digits_round))]

GR_merge_dt_agg$Bottom25= predict(loess(Bottom25~dMML,GR_merge_dt_agg),newdata=GR_merge_dt_agg$dMML)
GR_merge_dt_agg$top25= predict(loess(top25~dMML,GR_merge_dt_agg),newdata=GR_merge_dt_agg$dMML)
###plotting
pdf('../downstream/output/graphs/FigureS1/dNME_vs_dMML_quantile_differential_median.pdf',width=5,height=5)
print(ggplot(GR_merge_dt_agg,aes(x=dMML, y=dNME))+
        xlim(c(0,1))+ylim(c(0,0.7))+ggtitle("dMML and dNME relationship")+geom_smooth(method="loess",se=FALSE)+
        ylab("dNME")+theme(plot.title = element_text(hjust = 0.5),legend.position="bottom")+
        geom_ribbon(aes(ymin = Bottom25, ymax = top25), fill = "blue", alpha = .1)+theme_glob+
        scale_linetype_manual(values=c("solid","twodash", "twodash"))+scale_color_manual(values=c("Blue","Blue","Blue")))
dev.off()

# Plotting dNME vs dMML and NME vs MML ------------------------------------
#Figure S1B: MML and NME
GR_merge_dt=rbind(data.table(MML=GR_merge$MML1,NME=GR_merge$NME1,dNME_pval=GR_merge$dNME_pval,dMML_pval=GR_merge$dMML_pval),
                  data.table(MML=GR_merge$MML2,NME=GR_merge$NME2,dNME_pval=GR_merge$dNME_pval,dMML_pval=GR_merge$dMML_pval))
#Aggregate NME, using quantiles, 0.05 and 0.95
GR_merge_dt_agg=GR_merge_dt[, list(NME=round(median(NME),digits=digits_round),
                                   Bottom25=round(quantile(NME,probs=0.25),digits=digits_round),
                                   top25=round(quantile(NME,probs=0.75),digits=digits_round)), 
                            by = list(MML = round(MML,digits=digits_round))]

GR_merge_dt_agg$Bottom25= predict(loess(Bottom25~MML,GR_merge_dt_agg),newdata=GR_merge_dt_agg$MML)
GR_merge_dt_agg$top25= predict(loess(top25~MML,GR_merge_dt_agg),newdata=GR_merge_dt_agg$MML)
pdf('../downstream/output/graphs/FigureS1/NME_vs_MML_with_quantile.pdf',width=5,height=5)
#Plotting
print(ggplot(GR_merge_dt_agg,aes(x=MML, y=NME))+
        xlim(c(0,1))+ylim(c(0,1))+ggtitle("MML and NME relationship")+geom_smooth(method="loess",se=FALSE)+
        ylab("NME")+geom_ribbon(aes(ymin = Bottom25, ymax = top25), fill = "blue", alpha = .1)+theme_glob)
dev.off()


# pdf('../downstream/output/graphs/Figure2/imprinted_2A_density.pdf',width=3.5,height=3.5)
#  print(ggplot(GR_merge_genes_df,aes(x=dMML,fill=imprinted,color=imprinted))+geom_density(size=1.5,alpha=0.4)+
#        xlab('dMML')+theme_glob+theme(legend.position = 'bottom'))
# dev.off()


# ASM_SNP=sum(tb_in[[SNP_name]]==SNP&tb_in$dNME_pval<=pval_cutoff)
# ASM_nonSNP=sum(tb_in[[SNP_name]]!=SNP&tb_in$dNME_pval<=pval_cutoff)
# nonASM_SNP=sum(tb_in[[SNP_name]]==SNP&tb_in$dNME_pval>pval_cutoff)
# nonASM_nonSNP=sum(tb_in[[SNP_name]]!=SNP&tb_in$dNME_pval>pval_cutoff)
# OR=fisher.test(matrix(c(ASM_SNP,ASM_nonSNP,nonASM_SNP,nonASM_nonSNP),nrow=2))

# variant_meta$SNP=apply(variant_meta[,list(REF,ALT)],1,function(x) paste(x,collapse = '>'))
# SNP_freq=variant_meta[,list(count=length(REF)),by=list(SNP)]
# SNP_freq$freq=SNP_freq$count/sum(SNP_freq$count)
# pdf('../downstream/output/graphs/Figure3/SNP_frequency2.pdf',width=7,height=7)
# 
# ggplot(SNP_freq,aes(x=SNP,y=freq,fill=SNP))+geom_bar(stat="identity")+ylab('Frequency')+theme_glob+theme(legend.position = "bottom")
# dev.off()

# variant_SNP_OR=data.table()
# for (sn in unique(variant_meta$SNP)){
#   variant_SNP_OR=rbind(variant_SNP_OR,OR_calc(variant_meta,sn,"SNP",pval_cutoff ))
# }
# variant_SNP_OR=variant_SNP_OR[order(OR,decreasing = T)]
# variant_SNP_OR$SNP=factor(variant_SNP_OR$SNP,levels=variant_SNP_OR$SNP)
# variant_SNP_OR$FDR=p.adjust(variant_SNP_OR$pvalue,method="BH")
# variant_SNP_OR$sig=""
# variant_SNP_OR[FDR<=0.1]$sig="*"
# pdf('../downstream/output/graphs/Figure3/variant_OR2.pdf',width=7,height=7)
# ggplot(variant_SNP_OR,aes(x=SNP,y=OR,fill=SNP))+geom_bar(stat="identity")+ylab('OR')+theme_glob+theme(legend.position = "bottom")+
#   geom_errorbar(aes(ymin=lowerCI, ymax=upperCI), width=.2,position=position_dodge(.9))+ylim(c(0,1.5))+geom_text(aes(label=sig,y=upperCI),vjust=-0.5)
# dev.off()
# variant_meta$NME_relative=variant_meta$refNME-variant_meta$altNME
# pdf('../downstream/output/graphs/Figure3/variant_box.pdf',width=7,height=7)
# ggplot(variant_meta[dNME_pval<=pval_cutoff],aes(x=SNP,y=NME_relative,fill=SNP))+geom_boxplot()+ylab('refNME-altNME')+theme_glob+theme(legend.position = "bottom")+
#  ylim(c(-1,1))
# dev.off()

plot(GR_merge_tb[dMML_pval<=0.1]$NME,GR_merge_tb[dMML_pval<=0.1]$NME_ASM)

cor(GR_merge_tb[dMML_pval>pval_cutoff]$NME,GR_merge_tb[dMML_pval>pval_cutoff]$NME_ASM)
#Difference comes from coverage cutoff
get_density <- function(x, y, ...) {
  dens <- MASS::kde2d(x, y, ...)
  ix <- findInterval(x, dens$x)
  iy <- findInterval(y, dens$y)
  ii <- cbind(ix, iy)
  return(dens$z[ii])
}
library(viridis)
library(MASS)
GR_merge_gr_score_comp$density <- get_density(NME_in_score_comp$NME_agnostic , NME_in_score_comp$NME_allele_specific, n = 1000)
ggplot(NME_in_score_comp,aes(x=NME_allele_specific,y=NME_agnostic,color=density))+geom_point()+geom_smooth()+
  scale_color_viridis()+xlab("Allele-specific mean NME")+ylab("Allele-agnostic NME")
NME_in_score_comp$NME_diff=abs(NME_in_score_comp$NME_agnostic-NME_in_score_comp$NME_allele_specific)
NME_in_score_comp$density_diff <- get_density(NME_in_score_comp$NME_diff , NME_in_score_comp$dMML, n = 1000)
ggplot(NME_in_score_comp,aes(x=dMML,y=NME_diff))+geom_point()+geom_smooth()+
  scale_color_viridis()+xlab("dMML")+ylab("NME difference")
NME_in_score_comp_dMML=NME_in_score_comp[dMML_pval<=0.1]
NME_in_score_comp_dMML=rbind(NME_in_score_comp_dMML,NME_in_score_comp[dMML_pval>0.1][sample(1:sum(dMML_pval>0.1),133,replace = F)])
NME_in_score_comp_dMML$dMML_ASM=FALSE
NME_in_score_comp_dMML$dMML_ASM[NME_in_score_comp_dMML$dMML_pval<=0.1]=TRUE
NME_in_score_comp_dMML$dMML_ASM=as.factor(NME_in_score_comp_dMML$dMML_ASM)
ggplot(NME_in_score_comp_dMML,aes(x=NME_agnostic,group=dMML_ASM,colour=dMML_ASM))+geom_density()+
  xlab("allele agnostic NME")+theme(legend.position = "bottom")
GR_merge_tb=readRDS('../downstream/output/GR_merge_ASM_comp.rds')
GR_merge_tb=GR_merge_tb[!is.na(MML_ASM)&!is.na(MML)&!is.na(NME)]#3284912/52263042,total ASM regions=3332744
pdf('../downstream/output/graphs/FigureS4/boxplot_agnostic.pdf',height=3,width=3)
ggplot(GR_merge_tb,aes(y=NME,group=dMML_pval<=pval_cutoff,fill=dMML_pval<=pval_cutoff))+geom_boxplot()+theme_glob+
  theme(legend.position="bottom")+scale_fill_discrete(name = "dMML-ASM")
dev.off()
pdf('../downstream/output/graphs/FigureS4/boxplot_ASM.pdf',height=3,width=3)
ggplot(GR_merge_tb,aes(y=NME_ASM,group=dMML_pval<=pval_cutoff,fill=dMML_pval<=pval_cutoff))+geom_boxplot()+theme_glob+
  theme(legend.position="bottom")+scale_fill_discrete(name = "dMML-ASM")+ylab('NME')
dev.off()
# pdf('../downstream/output/graphs/FigureS4/boxplot_non_ASM.pdf',height=3,width=3)
# ggplot(GR_merge_tb,aes(y=NME,group=dMML_pval<=pval_cutoff,fill=dMML_pval<=pval_cutoff))+geom_boxplot()+theme_glob+
#   theme(legend.position="bottom")+scale_fill_discrete(name = "dMML-ASM")
# dev.off()
# pdf('../downstream/output/graphs/FigureS4/boxplot_ASM.pdf',height=3,width=3)
# ggplot(GR_merge_tb,aes(y=NME_ASM,group=dMML_pval<=pval_cutoff,fill=dMML_pval<=pval_cutoff))+geom_boxplot()+theme_glob+
#   theme(legend.position="bottom")+scale_fill_discrete(name = "dMML-ASM")+ylab('NME')
# dev.off()
wilcox.test(GR_merge_tb[dMML_pval<=pval_cutoff]$NME_diff,GR_merge_tb[dMML_pval>pval_cutoff]$NME_diff,alternative = 'greater')


plot_CpG_number<-function(allele_gr_ASM,stat="NME",theme_in=theme_glob){
  diff_pos=allele_gr_ASM$CpGdiff>0
  diff_neg=allele_gr_ASM$CpGdiff<0
  allele_het_df=data.table(More_CpG=c(elementMetadata(allele_gr_ASM)[,paste0(stat,'1')][diff_pos],
                                      elementMetadata(allele_gr_ASM)[,paste0(stat,'2')][diff_neg]),
                           less_CpG=c(elementMetadata(allele_gr_ASM)[,paste0(stat,'2')][diff_pos],
                                      elementMetadata(allele_gr_ASM)[,paste0(stat,'1')][diff_neg]),
                           N=c(allele_gr_ASM$N[diff_pos],allele_gr_ASM$N[diff_neg]))
  print(allele_het_df)
  print(ggplot(allele_het_df,aes(x=More_CpG,y=less_CpG))+
          # geom_bin2d(bins=100)+xlab("More CpG")+ylab("Less CpG")+
          geom_point(alpha=0.05)+xlab("More CpG")+ylab("Less CpG")+
          
          theme_in+theme(legend.position="bottom",legend.title = element_blank())+ylim(c(0,1)))
  #geom_abline(slope=1, colour="black", size=1))
}
# #3D debug:
# Fig3D=as.data.table(mcols(hyper_var_all$NME_hypervar_calc))
# Fig3D=Fig3D[!is.na(hypervar_logvar)]
# Fig3D=Fig3D[abs(dist)<=500]
# Fig3D=Fig3D[grepl("Spleen",Sample)]
# Fig3D_gene_hypervar_logvar=data.table(gene=Fig3D$gene,hypervar_logvar=Fig3D$hypervar_logvar)
# Fig3D_gene_hypervar_logvar=unique(Fig3D_gene_hypervar_logvar)
# Fig3D_gene_hypervar_logvar=quantile(Fig3D_gene_hypervar_logvar$hypervar_logvar,prob=seq(0.01,1,0.01),na.rm=T)
# Fig3D$quant001=findInterval(Fig3D$hypervar_logvar,Fig3D_gene_hypervar_logvar)/100
# Fig3D=Fig3D[,list(NME=median(score)),by=list(quant001,Sample)]
# print(ggplot(Fig3D,aes(quant001,Sample,fill=NME))+geom_tile()+scale_fill_distiller(palette = "RdPu", direction = 1)+
#         xlab('Hypervaribility quantile')+ylab('Sample')+theme_glob+theme(legend.position = 'bottom'))
# #Figure3D
# NME_hypervar_calc=hyper_var_all$NME_hypervar_calc[abs(hyper_var_all$NME_hypervar_calc$dist)<=500]
# NME_hypervar_calc=data.table(score=NME_hypervar_calc$score,hypervarquant=NME_hypervar_calc$hypervarquant001,Sample=NME_hypervar_calc$Sample)
# NME_hypervar_calc=NME_hypervar_calc[,list(median_score=median(score)),by=list(hypervarquant,Sample)]
# NME_hypervar_calc=NME_hypervar_calc[,list(hypervarquant=hypervarquant,NME=median_score,cor=cor(median_score,hypervarquant)),by=list(Sample)]
# NME_hypervar_calc=NME_hypervar_calc[order(NME_hypervar_calc$cor,decreasing = F),]
# NME_hypervar_calc$Sample=factor(NME_hypervar_calc$Sample,levels = unique(NME_hypervar_calc$Sample))
# pdf('../downstream/output/graphs/Figure3/Figure3B_hypervaribility_NME.pdf',width=7,height=7)
# print(ggplot(NME_hypervar_calc,aes(hypervarquant,Sample,fill=NME))+geom_tile()+scale_fill_distiller(palette = "RdPu", direction = 1)+
#   xlab('Hypervaribility quantile')+ylab('Sample')+theme_glob+theme(legend.position = 'bottom'))
# dev.off()



# GR_merge=readRDS(GR_merge_file)
# sample_hyper_var=unique(mcols(GR_merge)[,c('Sample','hyper_var_fn')])
# GR_merge_gr=convert_GR(GR_merge_tb$region)       
# mcols(GR_merge_gr)=GR_merge_tb
# GR_merge_gr$region=NULL
# NME_hypervar_calc=GRanges()
# MML_hypervar_calc=GRanges()
# NME_meanvar_calc=GRanges()
# MML_meanvar_calc=GRanges()
# GR_merge_gr$hypervar_logvar=NA
# genomic_features=readRDS(genomic_features_file)
# GR_merge_gr=dist_calc(GR_merge_gr,genomic_features$TSS)
# for (sp in unique(GR_merge_gr$Sample)){
#   
#   hyper_var_file=unlist(strsplit(sample_hyper_var$hyper_var_fn[sample_hyper_var$Sample==sp],';'))
#   cat('Processing',sp,'\n')
#   if(all(file.exists(hyper_var_file))){
#     
#     sp_hyper_var=read_hypervar(hyper_var_file)
#     #scRNA_result=rbind(scRNA_result,sp_hyper_var)
#     GR_merge_gr$hypervar_logvar[GR_merge_gr$Sample==sp]=
#       sp_hyper_var$hypervar_logvar[match(GR_merge_gr$gene[GR_merge_gr$Sample==sp],sp_hyper_var$gene_name)]
#  
#     
#   }else{cat("file not exist for:",sp,'\n')}
# }
# #Corelation between dMML and hypervaribility
# hyper_var_all_NME=hyper_var_all$NME_hypervar_calc
# print(cor.test(hyper_var_all_NME$dMML[abs(hyper_var_all_NME$dist)<=500],hyper_var_all_NME$exp_stat[abs(hyper_var_all_NME$dist)<=500]))
# NME_dMML_dt=data.table(NME=hyper_var_all_NME$score_original,dMML_pval=hyper_var_all_NME$dMML_pval,hyper_var=hyper_var_all_NME$exp_stat,dist=hyper_var_all_NME$dist)
# NME_dMML_dt$MML_ASM="non-dMML_ASM"
# NME_dMML_dt$MML_ASM[NME_dMML_dt$dMML_pval<=pval_cutoff]="dMML_ASM"
# pdf('../downstream/output/graphs/FigureS4/NME_dMML.pdf',width=3.5,height=3.5)
# print(ggplot(data=NME_dMML_dt,aes(x=NME,group=MML_ASM,color=MML_ASM))+
#   geom_density(size=1)+theme_glob+theme(legend.position="bottom",legend.title = element_blank()))
# dev.off()
# print(cor.test(NME_dMML_dt[abs(dist)<=500&MML_ASM=="dMML_ASM"]$NME,NME_dMML_dt[abs(dist)<=500&MML_ASM=="dMML_ASM"]$exp_stat))
#

# Plotting GO terms as example --------------------------------------------

plot_GO_example<-function(GO_in,fill){
  GO_in=GO_in[order(GO_in$FDR,-GO_in$FC)][1:5]
  #reverse order since bar plot plot bottom to top
  GO_in=GO_in[order(-GO_in$FDR,GO_in$FC)]
  GO_in$Term=factor(GO_in$Term,levels=GO_in$Term)
  print(ggplot(GO_in,aes(x=Term,y=-log10(FDR)))+geom_bar(stat="identity",fill=fill)+theme_glob+ coord_flip()+
          theme(legend.position = "",axis.text.y = element_blank(),axis.title.y = element_blank(),axis.title.x=element_text(size=8), axis.text.x = element_text(size=6))+
          geom_text(label=GO_in$Term,aes(y=0.1), hjust = 0,size=2.5))+xlab('-log10(P-value)')
  
}
cluster_col<- brewer.pal(10,'Set3')
#Plotting top10
mainDir='../downstream/output/graphs/Figure5/chromHMM_enhancer/'
GO_dir='../downstream/output/mm10_result/chromHMM_enhancer/cluster_GO/'
for(tissue in names(UC)){
  subDir=tissue
  ifelse(!dir.exists(file.path(mainDir, subDir)), dir.create(file.path(mainDir, subDir)), FALSE)
  for (clu in 1:10){
    clu_in=fread(paste0(GO_dir,tissue,'-',clu,'_cluster_GO.csv'))
    if(nrow(clu_in)>0){
      pdf(paste0(mainDir, subDir,'/',tissue,'-',clu,'.pdf'),width=1.5,height=1.5)
      print(plot_GO_example(clu_in,cluster_col[clu]))
      dev.off()
    }
  }
  
}
# Find top dNME genes for each cluster ------------------------------------
for (fn in dir('../downstream/output/mm10_result/chromHMM_enhancer/cluster_GO/')){
  GO_in=fread(paste0('../downstream/output/mm10_result/chromHMM_enhancer/cluster_GO/',fn))
  sp=sub('_.*','',fn)
  sp=sub("-",'_',sp)
  if(nrow(GO_in)>0){
    gene_GO=unique(unlist(strsplit(GO_in$genes,';')))
    gene_anno=fread(paste0('../downstream/input/mm10_cluster/',sp,'.csv'))
    gene_anno=gene_anno[gene_anno$chromHMM_enhancer&gene_anno$gene%in%gene_GO]
    gene_anno=gene_anno[,list(region=region[which.max(dNME_maxJSD)],dNME_maxJSD=dNME_maxJSD[which.max(dNME_maxJSD)],
                              distance=distance[which.max(dNME_maxJSD)],dMML_maxJSD=dMML_maxJSD[which.max(dNME_maxJSD)],
                              dMML_maxJSD_rank=dMML_maxJSD_rank[which.max(dNME_maxJSD)],
                              dNME_maxJSD_rank=dNME_maxJSD_rank[which.max(dNME_maxJSD)],
                              dNME_maxpair=dNME_maxpair[which.max(dNME_maxJSD)],
                              dMML_maxpair=dMML_maxpair[which.max(dNME_maxJSD)]),by=list(gene)]
    gene_anno=gene_anno[order(gene_anno$dNME_maxJSD,decreasing = T)]
    write.csv(gene_anno,paste0('../downstream/output/mm10_result/chromHMM_enhancer/cluster_gene/',sp,'_gene.csv'))
  }
}

# Find target gene in each motif that potentially be GO target ------------
motif_prefer_ent=fread('../downstream/output/graphs/tableS1_motif_prefer_ent_OMIM.csv')
motif_dir='../downstream/input/mouse_motif_cluster/'
motif_out=list()
for(tissue in dir(motif_dir)){
  motif_ts=data.table()
  for(fn in dir(paste0(motif_dir,tissue,'/'))){
    clu=strsplit(fn,'_')[[1]][4]
    motif_in=fread(paste0(motif_dir,tissue,'/',fn))
    motif_in$motif_short=sub('.*_','',motif_in$motif)
    motif_in=motif_in[FDR<=0.1&motif_short%in%motif_prefer_ent$TF]
    motif_in$cluster=clu
    motif_in$tissue=tissue
    motif_ts=rbind(motif_ts,motif_in)
  }
  motif_out[[tissue]]=motif_ts
}

saveRDS(motif_out,'../downstream/output/motif_mouse_overlap_human.rds')

JASPAR_mouse=readRDS('../downstream/')

for(tissue in names(motif_out)){
  
  
  
}

motif_dir='../downstream/input/Ken_list/'
GO_dir='../downstream/output/mm10_result/chromHMM_enhancer/'
for (fn in dir(motif_dir)){
  fn_split=strsplit(fn,'_')[[1]]
  tissue=fn_split[1]
  clu=fn_split[3]
  motif_name=fn_split[5]
  GO_gene=fread(paste0(GO_dir,'all_gene_list/',tissue,'_all.csv'))
  GO_gene=GO_gene[cluster==clu]
  GO_gene=GO_gene[order(dNME_maxJSD_rank,decreasing = F)]
  motif_in=fread(paste0(motif_dir,fn))
  gene_motif=GO_gene[region%in%paste0(motif_in$seqnames,':',motif_in$start,'-',motif_in$end)]
  write.csv(gene_motif[dNME_maxJSD>=0.3],paste0(GO_dir,'motif_target/',motif_name,'_',tissue,'_',clu,'_target.csv'))
}

# Adding promoter to the csv file -----------------------------------------
for(fn in dir('../downstream/input/mm10_cluster/')){
  csv_in=data.table()
  csv_in=fread(paste0('../downstream/input/mm10_cluster/',fn))
  csv_in_olap=as.data.table(findOverlaps(convert_GR(csv_in$region),promoters))
  csv_in_olap$gene=promoters$gene_name[csv_in_olap$subjectHits]
  csv_in_olap=csv_in_olap[,list(paste(gene,collapse = ';')),by=list(queryHits)]
  csv_in_olap=csv_in_olap[order(csv_in_olap$queryHits)]
  csv_in$promoter=FALSE
  csv_in$promoter[csv_in_olap$queryHits]=TRUE
  csv_in$promoter_gene=NA
  csv_in$promoter_gene[csv_in_olap$queryHits]=csv_in_olap$V1
  write.csv(csv_in,paste0('../downstream/input/mm10_cluster/',sub('.csv','_promoter.csv',fn)))
}

GO_run_tissue<-function(tissue,dir_in,nme_cor,mml_cor,enc_type){
  GO_out_all=list()
  csv_files=dir(paste0('../downstream/input/',dir_in),pattern="csv")
  for (ts in tissue){
    cat("Processing:",ts,'\n')
    fn=paste0(ts,'.csv')
    #read in csv file for given tissue
    csv_in_ts=fread(paste0('../downstream/input/',dir_in,'/',fn))
    csv_in_ts=csv_in_ts[order(dNME_maxJSD_rank,decreasing = F)]
    # Getting enhancer
    if(enc_type=="chromHMM_enhancer"){csv_in_ts=csv_in_ts[csv_in_ts$chromHMM_enhancer]}else
      if(enc_type=="non_chromHMM_enhancer"){csv_in_ts=csv_in_ts[!csv_in_ts$chromHMM_enhancer]}else 
        if(enc_type=="all_regions"){csv_in_ts=csv_in_ts}
    #GO annotation
    if(nrow(csv_in_ts)>1){
      #GO annotation for each cluster
      csv_out=lapply(1:10,function(clu){
        sp=paste0(ts,'-',clu)
        csv_in_ts_clu=csv_in_ts[cluster==clu]
        csv_in_ts_clu=csv_in_ts_clu[order(dNME_maxJSD_rank,decreasing=F)]
        #Add NME and mml cor
        csv_in_ts_clu$nme_cor=nme_cor[[ts]][match(csv_in_ts_clu$region,names(nme_cor[[ts]]))]
        csv_in_ts_clu$mml_cor=mml_cor[[ts]][match(csv_in_ts_clu$region,names(mml_cor[[ts]]))]
        
        if(nrow(csv_in_ts_clu)>1){
          
          #GO annotation for chromHMM 
          GO_out_cluster=GO_run(csv_in_ts_clu$gene,unique(csv_in_ts$gene),cluster=clu)
          csv_in_ts_clu$GO_result=unlist(lapply(unique(csv_in_ts_clu$gene),function(x) paste(GO_out_cluster$Term[grepl(x,GO_out_cluster$genes)],collapse = ';')))
          #GO annotation for other types of regions
          GO_out_cluster_NME=GO_run(csv_in_ts_clu[nme_cor>=0.7]$gene,unique(csv_in_ts$gene),cluster=clu)
          GO_out_cluster_NME_only=GO_run(csv_in_ts_clu[nme_cor>=0.7&mml_cor<0.7]$gene,unique(csv_in_ts$gene),cluster=clu)
          GO_out_cluster_MML=GO_run(csv_in_ts_clu[mml_cor>=0.7]$gene,unique(csv_in_ts$gene),cluster=clu)
          GO_out_cluster_MML_only=GO_run(csv_in_ts_clu[mml_cor>=0.7&nme_cor<0.7]$gene,unique(csv_in_ts$gene),cluster=clu)
          GO_out_cluster_non_MML=GO_run(csv_in_ts_clu[mml_cor<0.7]$gene,unique(csv_in_ts$gene),cluster=clu)
          GO_out_cluster_NME_MML=GO_run(csv_in_ts_clu[nme_cor>=0.7&mml_cor>=0.7]$gene,unique(csv_in_ts$gene),cluster=clu)
          GO_out_cluster_non_NME_non_MML=GO_run(csv_in_ts_clu[nme_cor<0.7&mml_cor<0.7]$gene,unique(csv_in_ts$gene),cluster=clu)
          
          write.csv(GO_out_cluster,row.names = F,quote = T,
                    file=paste0('../downstream/output/mm10_result/',enc_type,'/cluster_GO/',dir_in,'/all_regions/',sp,'_cluster_GO.csv'))
          write.csv(GO_out_cluster_NME_only,row.names = F,quote = T,
                    file=paste0('../downstream/output/mm10_result/',enc_type,'/cluster_GO/',dir_in,'/NME_only/',sp,'_cluster_GO.csv'))
          write.csv(GO_out_cluster_MML_only,row.names = F,quote = T,
                    file=paste0('../downstream/output/mm10_result/',enc_type,'/cluster_GO/',dir_in,'/MML_only/',sp,'_cluster_GO.csv'))
          write.csv(GO_out_cluster_NME_MML,row.names = F,quote = T,
                    file=paste0('../downstream/output/mm10_result/',enc_type,'/cluster_GO/',dir_in,'/NME_MML/',sp,'_cluster_GO.csv'))
          write.csv(GO_out_cluster_non_NME_non_MML,row.names = F,quote = T,
                    file=paste0('../downstream/output/mm10_result/',enc_type,'/cluster_GO/',dir_in,'/None_NME_MML/',sp,'_cluster_GO.csv'))
          #return(list(csv_in_ts_clu=csv_in_ts_clu,GO_out_cluster=GO_out_cluster))
          return(list(GO_out_cluster_all=GO_out_cluster,GO_out_cluster_NME=GO_out_cluster_NME,GO_out_cluster_non_MML=GO_out_cluster_non_MML,csv_in_ts_clu=csv_in_ts_clu,
                      GO_out_cluster_NME_only=GO_out_cluster_NME_only,GO_out_cluster_MML=GO_out_cluster_MML,GO_out_cluster_MML_only=GO_out_cluster_MML_only,
                      GO_out_cluster_non_NME_non_MML=GO_out_cluster_non_NME_non_MML,GO_out_cluster_NME_MML=GO_out_cluster_NME_MML))
        }
        
      })
      GO_out_all[[ts]]=csv_out
      write.csv(fastDoCall('rbind',lapply(csv_out,function(x) x$csv_in_ts_clu))[order(dNME_maxJSD,decreasing=T)],
                file=paste0('../downstream/output/mm10_result/',enc_type,'/enhancer_gene_list/',dir_in,'/',ts,'_all.csv'))
    }
    
  }
  return(GO_out_all)
}

GO_run_tissue_perm<-function(tissue,dir_in,enc_type,dist_cutoff=NA,permute=F){
  if(permute){enc_type=paste0('permute',enc_type)}
  GO_out_all=list()
  # dir_out=paste0('../downstream/output/mm10_result/',dir_in)
  # ifelse(!dir.exists(file.path(dir_out)), dir.create(file.path(dir_out)), FALSE)
  # dir_out=paste0(dir_out,'/',enc_type,'_',dist_cutoff)
  # ifelse(!dir.exists(file.path(dir_out)), dir.create(file.path(dir_out)), FALSE)
  # GO_out=paste0(dir_out,'/cluster_GO/')
  # ifelse(!dir.exists(file.path(GO_out)), dir.create(file.path(GO_out)), FALSE)
  # gene_out=paste0(dir_out,'/all_gene_list/')
  # ifelse(!dir.exists(file.path(gene_out)), dir.create(file.path(gene_out)), FALSE)
  
  csv_files=dir(paste0('../downstream/input/',dir_in),pattern="csv")
  print(csv_files)
  for (ts in tissue){
    cat("Processing:",ts,'\n')
    fn=paste0(ts,'.csv')
    #read in csv file for given tissue
    csv_in_ts=fread(paste0('../downstream/input/',dir_in,'/',fn))
    
    #Note some times Jason use dNME_maxJSD_rank
    #csv_in_ts=csv_in_ts[order(dNME_maxUC_rank,decreasing = F)]
    # Getting enhancer
    if(enc_type=="chromHMM_enhancer"){csv_in_ts=csv_in_ts[csv_in_ts$chromHMM_enhancer]}else
      if(enc_type=="non_chromHMM_enhancer"){csv_in_ts=csv_in_ts[!csv_in_ts$chromHMM_enhancer]}else 
        if(enc_type=="promoter"){csv_in_ts=csv_in_ts}else 
          if(enc_type=="all_regions"){csv_in_ts=csv_in_ts}else
            if(enc_type=="bin_enhancer"){
              enhancer=readRDS("../downstream/output/bin_enhancer.rds")
              csv_in_gr=convert_GR(csv_in_ts$region)
              mcols(csv_in_gr)=csv_in_ts
              olap=findOverlaps(csv_in_gr,enhancer)
              csv_in_gr=csv_in_gr[queryHits(olap)]
              csv_in_gr$gene=enhancer$`Target Gene`[subjectHits(olap)]
              csv_in_gr$distance=NA
              csv_in_ts=as.data.table(mcols(csv_in_gr))
            }else
              if(enc_type=="FANTOM5"){
                #heart 47831 regions
                enhancer=import.bed('../downstream/input/F5.mm10.enhancers.bed.gz')
                csv_in_gr=convert_GR(csv_in_ts$region)
                olap=findOverlaps(csv_in_gr,enhancer)
                print("Filtering")
                csv_in_ts=csv_in_ts[queryHits(olap)]
              }else
                if(enc_type=="chromHMM_enhancer_union"){
                  csv_in_ts=csv_in_ts[csv_in_ts$chromHMM_enhancer_union]
                  
                }else
                  if(enc_type=="chromHMM_enhancer_tissuespecific"){
                    csv_in_ts=csv_in_ts[csv_in_ts$chromHMM_enhancer_tissuespecific]
                    
                  }
    
    
    if(!is.na(dist_cutoff)){csv_in_ts=csv_in_ts[abs(distance)<=dist_cutoff]}
    if(permute){csv_in_ts$cluster=sample(csv_in_ts$cluster,length(csv_in_ts$cluster))}
    print(csv_in_ts)
    #GO annotation
    if(nrow(csv_in_ts)>1){
      #GO annotation for each cluster
      csv_out=mclapply(1:10,function(clu){
        sp=paste0(ts,'-',clu)
        csv_in_ts_clu=csv_in_ts[cluster==clu]
        #csv_in_ts_clu=csv_in_ts_clu[order(dNME_maxUC_rank,decreasing=F)]
        #Add NME and mml cor
        # csv_in_ts_clu$nme_cor=nme_cor[[ts]][match(csv_in_ts_clu$region,names(nme_cor[[ts]]))]
        # csv_in_ts_clu$mml_cor=mml_cor[[ts]][match(csv_in_ts_clu$region,names(mml_cor[[ts]]))]
        
        if(nrow(csv_in_ts_clu)>1){
          Sys.sleep(clu*30)
          cat('start processing cluster:',clu,'\n')
          tt1=proc.time()[[3]]
          #GO annotation for chromHMM 
          GO_out_cluster=GO_run(csv_in_ts_clu$gene,unique(csv_in_ts$gene),cluster=clu)
          csv_in_ts_clu$GO_result=unlist(lapply(csv_in_ts_clu$gene,function(x) paste(GO_out_cluster$Term[grepl(x,GO_out_cluster$genes)],collapse = ';')))
          cat('Finish processing cluster:',clu,'in:',proc.time()[[3]]-tt1,'\n')
          
          # write.csv(GO_out_cluster,row.names = F,quote = T,
          #           file=paste0(GO_out,sp,'_cluster_GO.csv'))
          #return(list(csv_in_ts_clu=csv_in_ts_clu,GO_out_cluster=GO_out_cluster))
          return(list(GO_out_cluster_all=GO_out_cluster,csv_in_ts_clu=csv_in_ts_clu))
        }
        
      },mc.cores=10)
      print(csv_out[[1]])
      GO_out_all[[ts]]=csv_out
      # write.csv(fastDoCall('rbind',lapply(csv_out,function(x) x$csv_in_ts_clu))[order(dNME_maxUC,decreasing=T)],
      #           file=paste0(gene_out,ts,'_all.csv'))
    }
    
  }
  return(GO_out_all)
}
